tests/acceptance/test_external_studies.py

```diff
@@ -0,0 +1,88 @@
+import shutil
+import sys
+import tarfile
+from fnmatch import fnmatch
+from pathlib import Path
+from urllib.request import urlopen
+
+import pytest
+
+from databuilder.main import load_dataset_definition
+from databuilder.query_language import compile
+
+# These tests specify dataset definitions in other repositories which we want to ensure
+# we don't accidentally break. In order to keep tests hermetic and deterministic, we
+# copy the study code into the repo and commit it (taking care to copy just the files we
+# need to evalute the dataset definition). This module can also be used as a script
+# which ensures our vendored copy of the code is up-to-date. Usage is:
+#
+#     python -m tests.acceptance.test_external_studies update
+#
+# Or via just as:
+#
+#     just update-external-studies
+#
+# This is run automatically by a scheduled action which will create a PR if there are
+# any changes to be made.
+EXTERNAL_STUDIES = {
+    "test-age-distribution": dict(
+        repo="opensafely/test-age-distribution",
+        branch="main",
+        file_globs=[
+            "analysis/dataset_definition.py",
+        ],
+        dataset_definition="analysis/dataset_definition.py",
+    ),
+}
+
+STUDY_DIR = Path(__file__).parent / "external_studies"
+
+
+@pytest.mark.parametrize("name", EXTERNAL_STUDIES.keys())
+def test_external_study(name):
+    study_path = STUDY_DIR / name
+    dataset_def_path = study_path / EXTERNAL_STUDIES[name]["dataset_definition"]
+    # Test that we can compile the dataset definition to a valid query model graph. I
+    # think this is sufficient for these tests which are intended to ensure we don't
+    # accidentally break the API. If we're unable to execute a valid query, that's a
+    # separate class of problem for which we need separate tests.
+    dataset = load_dataset_definition(dataset_def_path)
+    variable_definitions = compile(dataset)
+    assert variable_definitions
+
+
+# SCRIPT TO UPDATE EXTERNAL STUDIES
```

Comment:
Feels like this should be in a separate file.

Reply:
Yeah, fair enough. I was trying to keep it self-contained but it does mix up the concerns a bit too much.